Approach.md

```diff
@@ -0,0 +1,160 @@
+
+The daemon will be run entirely in memory and will not persist any state in order to survive reboots/crashes.
+
+The gRPC API is assumed to be internal only, as such there won't be any rate limiting, user action logging or abuse mitigations.
+
+Dealing with tty, ptty, terminals and what not is entirely out of scope and as such any program output will be streamed as the raw bytes written to the stdout and stderr pipes.
+Furthermore the output will be chunked on newlines and any lines longer than max gRPC packet size will be truncated.
+
+Resource limits will be implemented using separate cgroups for each
+
+Isolation will be hardcoded to fully isolate each scheduled task from other tasks and the host system. That means a separate PID, Mount, User and Network namespace for each task with no way to connect them.
+There will be no network devices attached to any tasks.
+
+Mounting is also considered out of scope and as such each task will be chrooted to a RW root that's a copy of a base image that's removed upon exit of the task.
+
+I'll make no attempt at Linux backwards compatibility and the minimum kernel version will be 5.0+.
+
+Tasks will inherit the UID/GID of the daemon and internally be mapped as root.
+
+For the gRPC implementation I'll be using the rust crate [tonic](https://github.com/hyperium/tonic) to avoid reinventing the wheel.
+
+Mutable interaction with tasks will be mutually exclusive meaning that the stop action will take a write-only lock on the task such that any concurrent stop task will wait for the first to complete and the subsequently fail as the task has already stopped.
+Read-only interactions such as query and stream can happen concurrently without issue.
+
+Output streaming will be implemented by redirecting the spawned process' stdout and stderr to pipes created with `pipe(2)`.
+These pipes can then be read in an async fashion from a tokio reader task and passed on to a `broadcast` multiple-producer multiple-consumer (MPMC) channel (only a single producer will exists but there's no SPMC channel in tokio).
+Then from the gRPC server side it's easy to lookup the `broadcast` channel for a given task-id and subscribe to it in an async task which produces matching `TaskOutputReply` messages and writes them to the gRPC stream.
+Furthermore once the spawned process exit, the pipes will close which in turn will be detected by the reader task resulting in the channel being closed which last but not least will make the gRPC stream end with a final message containing the task status.
+This cuts one significant corner, namely it's not possible to retrieve past output which most certainly would be a requirement for a production system.
+For a production system you'd implement some kind of ring-buffer that enables fetching the past N bytes of output while subscribing to future output without falling behind.

```

Comment:
The streaming endpoint should actually start from the beginning of the process's output, even if the job has already exited.  No cleanup mechanism is required though.  Its OK for server memory to grow indefinitely as mode jobs are run.  Its more important that we have consistent and compete output.

Reply:
It wasn't clear to me from the specification that it's a requirement to output the full output and not just future output.
Please have a look at the newest commit, I've described how I'll implement full output streaming.